File: csharp/Fichas/Fichas.Lectura/Normalizacion.cs
using System.Globalization;
using System.Text.RegularExpressions;

namespace Fichas.Lectura;

/// <summary>
/// Convertir el texto leido en un valor con formato, o en nada.
/// </summary>
/// <remarks>
/// Portado de `extraccion/normalizacion.py`. Todo lo de aqui son expresiones regulares y
/// tablas. Ni una sola de estas funciones adivina: cuando el texto no encaja devuelven
/// nulo, y nulo significa «no se pudo leer», que es una respuesta correcta. La respuesta
/// incorrecta seria una cedula inventada.
///
/// <para>⛔ Regla permanente 1 del proyecto, aplicada donde mas tienta romperla.</para>
/// </remarks>
public static partial class Normalizacion
{
    // --- Los patrones ------------------------------------------------------------

    /// <summary>
    /// Cuatro letras MAYÚSCULAS y cuatro dígitos, como palabra entera. Solo mayúsculas a
    /// propósito: ver <see cref="NormalizarNumeroDeCaso"/>.
    /// </summary>
    [GeneratedRegex(@"\b([A-Z]{4}[0-9]{4})\b")]
    private static partial Regex PatronNumeroDeCaso();

    /// <summary>
    /// El ultimo caracter de la cedula puede ser una LETRA (`DECISIONES.md`, 2026-09-04).
    /// El patron viejo pedia once digitos en 3-4-4 y perdia 2 de las 7 cedulas de los
    /// escaneos reales del dueno: el OCR las leia bien y la validacion las tiraba.
    /// </summary>
    [GeneratedRegex(@"\b([0-9]{3})-([0-9]{4})-([0-9]{3}[0-9A-Za-z])\b")]
    private static partial Regex PatronCedula();

    /// <summary>
    /// Seis dígitos O MÁS seguidos. El «o más» no es descuido: captura la cifra entera para
    /// que <see cref="LargosDeUnidad"/> la rechace si es de 8, en vez de recortarla a 7.
    /// </summary>
    [GeneratedRegex(@"\b([0-9]{6,})\b")]
    private static partial Regex PatronUnidad();

    /// <summary>Uno o más blancos seguidos, para dejarlos en un solo espacio.</summary>
    [GeneratedRegex(@"\s+")]
    private static partial Regex EspaciosSeguidos();

    /// <summary>Año-mes-día con cuatro cifras de año delante, separado por guion, barra o punto: «2026-09-08».</summary>
    [GeneratedRegex(@"\b(\d{4})[-/.](\d{1,2})[-/.](\d{1,2})\b")]
    private static partial Regex FechaIso();

    /// <summary>
    /// Dos números de una o dos cifras y un año de cuatro: «08-09-2026». Cuál es el día y
    /// cuál el mes lo decide <see cref="DiaYMesSinAdivinar"/>, no el patrón.
    /// </summary>
    [GeneratedRegex(@"\b(\d{1,2})[-/.](\d{1,2})[-/.](\d{4})\b")]
    private static partial Regex FechaEnNumeros();

    /// <summary>
    /// Los largos que el numero de unidad admite, en un solo sitio.
    /// </summary>
    /// <remarks>
    /// 6 o 7 digitos: `DECISIONES.md` 2026-09-02, «manda el papel». El patron captura 6 o
    /// mas a proposito, para ver una cifra de 8 entera, reconocer que NO es un numero de
    /// unidad y devolver nulo, en vez de recortarle los dos ultimos digitos y dar por
    /// bueno un numero que no existe.
    /// </remarks>
    private static readonly int[] LargosDeUnidad = [6, 7];

    // --- Numero de caso ----------------------------------------------------------

    /// <summary>
    /// 4 letras mayusculas y 4 digitos, o nulo.
    /// </summary>
    /// <remarks>
    /// No se pasa a mayusculas lo que venia en minusculas: si el OCR leyo `casp` puede
    /// haber leido mal tambien las cifras, y forzarlo a `CASP` esconderia el problema en
    /// vez de mandarlo a revision.
    /// </remarks>
    /// <param name="texto">Lo leído o lo escrito; puede traer más cosas alrededor, se busca dentro.</param>
    /// <returns>El primer número de caso con forma que aparezca, o nulo.</returns>
    public static string? NormalizarNumeroDeCaso(string? texto)
    {
        if (string.IsNullOrWhiteSpace(texto)) return null;
        var coincidencia = PatronNumeroDeCaso().Match(texto);
        return coincidencia.Success ? coincidencia.Groups[1].Value : null;
    }

    // --- Cedula de miembro -------------------------------------------------------

    /// <summary>
    /// 3 digitos, guion, 4 digitos, guion, 3 digitos y un caracter mas. O nulo.
    /// </summary>
    /// <remarks>
    /// Ese ultimo caracter puede ser un digito o una LETRA: `055-1111-3853` y
    /// `066-2222-133A` son las dos FORMAS, y las dos salen de escaneos reales del dueno.
    /// Palabras suyas el 2026-09-04: «muchas cedulas de miembro tienen una A u otra letra
    /// al final».
    ///
    /// <para>⚠️ Los dos valores de arriba estan SUSTITUIDOS: lo que salio del papel fue la
    /// forma, y la forma esta entera; el ejemplar concreto no vive en el repositorio desde
    /// el 2026-09-07. Ver `EN-CURSO.md`, «Los datos de personas reales salen del
    /// repositorio». Vale para todos los ejemplos de cedula de este archivo.</para>
    ///
    /// <para>La letra se devuelve tal como venia: no se sube a mayuscula, no se cambia
    /// por un digito «parecido» y no se recorta. Cualquiera de las tres cosas es inventar
    /// un dato (regla permanente 1), y una cedula inventada manda a una persona al templo
    /// con la recomendacion equivocada.</para>
    /// </remarks>
    /// <param name="texto">Lo leído o lo escrito; puede traer el nombre u otras cosas alrededor.</param>
    /// <returns>La primera cédula con forma, reconstruida con sus dos guiones, o nulo.</returns>
    public static string? NormalizarCedula(string? texto)
    {
        if (string.IsNullOrWhiteSpace(texto)) return null;
        var coincidencia = PatronCedula().Match(texto);
        if (!coincidencia.Success) return null;

        return $"{coincidencia.Groups[1].Value}-{coincidencia.Groups[2].Value}-{coincidencia.Groups[3].Value}";
    }

    // --- Nombre de persona -------------------------------------------------------

    /// <summary>
    /// El nombre de la persona sin la cedula que el OCR le pego delante o detras.
    /// </summary>
    /// <remarks>
    /// <para><b>El caso medido</b>, en los siete escaneos del dueño: el OCR devolvio
    /// «Ejemplo, Daniel Jr. Damian Dorian |055-1111-3853 Verified √» como una sola caja.
    /// Cuando eso pasa, el nombre y la cedula viven en la misma linea, y sin quitarla el
    /// nombre de la persona entra a la base con su cedula dentro. Palabras del dueño el
    /// 2026-09-07: «siempre debe ser el nombre y debajo la cedula de miembro de la persona».
    /// </para>
    ///
    /// <para>⛔ <b>Esto NO es arreglar texto leido</b> (regla permanente 1). No se corrige
    /// ni una letra ni se adivina nada: se quita un trozo que tiene la forma exacta de una
    /// cedula, porque ese trozo se va a su propia columna. Es la misma operacion que
    /// <see cref="NormalizarUnidad"/> lleva haciendo desde el principio con el nombre y el
    /// numero del barrio, que tambien vienen pegados en una sola linea del papel.</para>
    ///
    /// <para>Se aplica SOLO al leer una fila del papel, no a lo que Miguel teclea: ver
    /// <c>Extraccion.NormalizadorDe</c>. Un nombre escrito a mano no trae una cedula dentro,
    /// y quitarle algo a lo que una persona escribio seria otra cosa.</para>
    /// </remarks>
    /// <param name="texto">El renglón entero de la persona tal como lo leyó el OCR.</param>
    /// <returns>El texto sin ninguna cédula con forma y sin los separadores que quedan en los bordes, o nulo si no queda nada.</returns>
    public static string? NombreSinLaCedula(string? texto)
    {
        if (string.IsNullOrWhiteSpace(texto)) return null;

        string sinCedula = PatronCedula().Replace(texto, " ");
        string junto = EspaciosSeguidos().Replace(sinCedula, " ").Trim(' ', '|', '-', '–', ',', '\t');
        return junto.Length == 0 ? null : junto;
    }

    // --- Unidad ------------------------------------------------------------------

    /// <summary>
    /// El numero de unidad y el resto del texto. Cualquiera de los dos puede ser nulo.
    /// </summary>
    /// <remarks>
    /// Una cifra de otro largo NO se recorta ni se rellena: vuelve nula y el texto crudo
    /// viaja aparte para que se vea que habia algo.
    /// </remarks>
    /// <param name="texto">La banda de la unidad entera: nombre y número vienen pegados en una línea.</param>
    /// <returns>El número si tiene 6 o 7 dígitos, y el resto del texto como nombre; sin cifra, todo el texto es el nombre.</returns>
    public static (string? Numero, string? Nombre) NormalizarUnidad(string? texto)
    {
        if (string.IsNullOrWhiteSpace(texto)) return (null, null);

        var coincidencia = PatronUnidad().Match(texto);
        if (!coincidencia.Success)
        {
            string suelto = texto.Trim();
            return (null, suelto.Length == 0 ? null : suelto);
        }

        string digitos = coincidencia.Groups[1].Value;
        string nombre = (texto[..coincidencia.Index] + texto[(coincidencia.Index + coincidencia.Length)..])
            .Trim(' ', '-', '–', '\t');
        string? numero = LargosDeUnidad.Contains(digitos.Length) ? digitos : null;

        return (numero, nombre.Length == 0 ? null : nombre);
    }

    /// <summary>Solo el numero de unidad, o nulo.</summary>
    /// <remarks>
    /// Existe aparte porque el numero y el nombre se corrigen por separado en los
    /// formularios reales: hay anotaciones que reescriben el nombre y dejan el numero
    /// como estaba.
    /// </remarks>
    /// <param name="texto">La banda de la unidad, o solo una cifra escrita a mano.</param>
    public static string? NormalizarNumeroDeUnidad(string? texto) => NormalizarUnidad(texto).Numero;

    /// <summary>Solo el nombre de la unidad, sin su numero, o nulo.</summary>
    /// <param name="texto">La banda de la unidad, o solo un nombre escrito a mano.</param>
    public static string? NormalizarNombreDeUnidad(string? texto) => NormalizarUnidad(texto).Nombre;

    // --- Templo ------------------------------------------------------------------

    /// <summary>
    /// El nombre del templo tal como esta escrito, con los espacios juntados.
    /// </summary>
    /// <remarks>
    /// <b>No hay catalogo y no se corrige nada</b> (regla permanente 1). No se compara
    /// contra una lista de templos ni se elige «el mas parecido»: el catalogo es una
    /// decision del dueno que todavia no ha tomado, y hasta entonces un templo mal leido
    /// tiene que verse mal leido y no convertido en otro que si esta en la lista.
    /// </remarks>
    /// <param name="texto">Lo leído en la banda del templo o lo escrito a mano.</param>
    /// <returns>El mismo texto recortado y con un solo espacio entre palabras, o nulo si estaba en blanco.</returns>
    public static string? NormalizarNombreDelTemplo(string? texto)
    {
        if (string.IsNullOrWhiteSpace(texto)) return null;
        string junto = EspaciosSeguidos().Replace(texto, " ").Trim();
        return junto.Length == 0 ? null : junto;
    }

    // --- Fecha -------------------------------------------------------------------

    /// <summary>
    /// Cada nombre o abreviatura de mes, en español y en inglés, con su número. Sin
    /// distinguir mayúsculas. «may» y «mar» sirven a los dos idiomas; «setiembre» está
    /// porque se escribe así en parte de América.
    /// </summary>
    private static readonly Dictionary<string, int> Meses = new(StringComparer.OrdinalIgnoreCase)
    {
        ["enero"] = 1, ["ene"] = 1, ["january"] = 1, ["jan"] = 1,
        ["febrero"] = 2, ["feb"] = 2, ["february"] = 2,
        ["marzo"] = 3, ["mar"] = 3, ["march"] = 3,
        ["abril"] = 4, ["abr"] = 4, ["april"] = 4, ["apr"] = 4,
        ["mayo"] = 5, ["may"] = 5,
        ["junio"] = 6, ["jun"] = 6, ["june"] = 6,
        ["julio"] = 7, ["jul"] = 7, ["july"] = 7,
        ["agosto"] = 8, ["ago"] = 8, ["august"] = 8, ["aug"] = 8,
        ["septiembre"] = 9, ["setiembre"] = 9, ["september"] = 9, ["sept"] = 9, ["sep"] = 9,
        ["octubre"] = 10, ["oct"] = 10, ["october"] = 10,
        ["noviembre"] = 11, ["nov"] = 11, ["november"] = 11,
        ["diciembre"] = 12, ["dic"] = 12, ["december"] = 12, ["dec"] = 12,
    };

    /// <summary>El sufijo ordinal inglés que puede seguir al día («2nd», «8th»), opcional. Se casa y se tira.</summary>
    private const string SufijoOrdinal = "(?:st|nd|rd|th)?";

    /// <summary>
    /// Las claves de <see cref="Meses"/> como alternativa de expresión regular, las más
    /// largas primero: si «sep» fuera antes que «september», el motor se quedaría con las
    /// tres letras y el resto de la palabra rompería el patrón.
    /// </summary>
    private static readonly string NombresDeMes =
        string.Join("|", Meses.Keys.OrderByDescending(nombre => nombre.Length));

    /// <summary>«September 8, 2026», «Sept-7, 2026», «September 2nd, 2026».</summary>
    private static readonly Regex MesDiaAnio = new(
        $@"\b({NombresDeMes})\.?\s*[-\s]\s*(\d{{1,2}}){SufijoOrdinal}\s*,?\s*(\d{{4}})\b",
        RegexOptions.IgnoreCase | RegexOptions.Compiled);

    /// <summary>«8 Sept 2026», «14 de marzo de 2027». Los dos «de» son opcionales: el OCR se come uno.</summary>
    private static readonly Regex DiaMesAnio = new(
        $@"\b(\d{{1,2}}){SufijoOrdinal}\s+(?:de\s+)?({NombresDeMes})\.?\s*(?:de\s+|,\s*)?(\d{{4}})\b",
        RegexOptions.IgnoreCase | RegexOptions.Compiled);

    /// <summary>
    /// Un rango —«September 8-11, 2026»— NO es una fecha de viaje: no se sabe cual de los
    /// dos dias vale. Se detecta para poder rechazarlo a proposito en vez de quedarse con
    /// el primero por casualidad. Es el texto real de la cita del templo de los siete.
    /// </summary>
    private static readonly Regex Rango = new(
        $@"\b\d{{1,2}}{SufijoOrdinal}\s*(?:-|–|to|till|until|through)\s*\d{{1,2}}{SufijoOrdinal}\b",
        RegexOptions.IgnoreCase | RegexOptions.Compiled);

    /// <summary>La fecha en ISO-8601, o nula si el dia no existe en ese mes.</summary>
    /// <param name="anio">De 1 a 9999; fuera de eso, nulo.</param>
    /// <param name="mes">De 1 a 12; fuera de eso, nulo.</param>
    /// <param name="dia">Se comprueba contra los días reales de ese mes y año, bisiestos incluidos.</param>
    /// <returns>«aaaa-MM-dd» en cultura invariante, o nulo. Nunca lanza por una fecha imposible.</returns>
    private static string? FechaIso8601(int anio, int mes, int dia)
    {
        if (mes < 1 || mes > 12 || anio < 1 || anio > 9999) return null;
        if (dia < 1 || dia > DateTime.DaysInMonth(anio, mes)) return null;
        return new DateOnly(anio, mes, dia).ToString("yyyy-MM-dd", CultureInfo.InvariantCulture);
    }

    /// <summary>
    /// Cual de los dos numeros es el dia y cual el mes, o nada.
    /// </summary>
    /// <remarks>
    /// Solo responde cuando los DIGITOS lo deciden: un numero mayor que 12 no puede ser
    /// un mes, asi que el otro lo es. Cuando los dos son 12 o menos la fecha es ambigua
    /// de verdad —«05-10-2027» es el 5 de octubre o el 10 de mayo— y esta funcion se
    /// niega a elegir. Suponer «dia-mes porque el dueno es de un pais que lo escribe asi»
    /// acertaria la mayoria de las veces; las que fallara mueven una fecha de viaje
    /// varios meses, y es la fecha que decide si un caso sale avisado a tiempo.
    /// </remarks>
    /// <param name="primero">El número que va antes del primer separador.</param>
    /// <param name="segundo">El número que va después.</param>
    /// <returns>El par (día, mes) cuando uno de los dos no puede ser mes; nulo cuando los dos podrían serlo.</returns>
    private static (int Dia, int Mes)? DiaYMesSinAdivinar(int primero, int segundo)
    {
        if (primero > 12 && segundo <= 12) return (primero, segundo);
        if (segundo > 12 && primero <= 12) return (segundo, primero);
        return null;
    }

    /// <summary>
    /// La fecha escrita en cifras, o nula si no hay exactamente UNA sin ambiguedad.
    /// </summary>
    /// <remarks>
    /// Se juntan todas las que aparezcan y solo vale si todas dicen lo mismo: dos fechas
    /// distintas en la misma banda son un rango, y un rango no dice cuando se viaja.
    /// <para>Una sola fecha ambigua en cifras («05-10-2027») anula el resultado entero,
    /// aunque al lado hubiera una en ISO clara: no se puede afirmar que las dos digan lo mismo.</para>
    /// </remarks>
    /// <param name="texto">La banda entera; no nulo, quien llama ya lo comprobó.</param>
    private static string? FechaEnCifras(string texto)
    {
        var encontradas = new HashSet<string>(StringComparer.Ordinal);

        foreach (Match m in FechaIso().Matches(texto))
        {
            var iso = FechaIso8601(int.Parse(m.Groups[1].Value), int.Parse(m.Groups[2].Value), int.Parse(m.Groups[3].Value));
            if (iso is not null) encontradas.Add(iso);
        }
        foreach (Match m in FechaEnNumeros().Matches(texto))
        {
            var decidido = DiaYMesSinAdivinar(int.Parse(m.Groups[1].Value), int.Parse(m.Groups[2].Value));
            if (decidido is null) return null;
            var iso = FechaIso8601(int.Parse(m.Groups[3].Value), decidido.Value.Mes, decidido.Value.Dia);
            if (iso is not null) encontradas.Add(iso);
        }
        return encontradas.Count == 1 ? encontradas.First() : null;
    }

    /// <summary>
    /// «8 Sept 2026» y «08-09-2026» dan «2026-09-08». Nula si no hay UNA fecha clara.
    /// </summary>
    /// <remarks>
    /// Las cifras se miran ANTES que el rango, y no es un detalle de orden: el detector
    /// de rangos casa con «14-03-2027» —dos numeros de dos cifras con un guion en medio—
    /// y sin esto se comeria todas las fechas del formulario espanol dandolas por rangos.
    /// </remarks>
    /// <param name="texto">La banda de la fecha o lo escrito a mano; puede traer texto alrededor.</param>
    /// <returns>«aaaa-MM-dd», o nulo si no hay fecha, hay un rango, la fecha es ambigua o el día no existe.</returns>
    public static string? NormalizarFecha(string? texto)
    {
        if (string.IsNullOrWhiteSpace(texto)) return null;

        var enCifras = FechaEnCifras(texto);
        if (enCifras is not null) return enCifras;
        if (Rango.IsMatch(texto)) return null;

        var conMesPrimero = MesDiaAnio.Match(texto);
        if (conMesPrimero.Success)
        {
            return FechaIso8601(
                int.Parse(conMesPrimero.Groups[3].Value),
                Meses[conMesPrimero.Groups[1].Value],
                int.Parse(conMesPrimero.Groups[2].Value));
        }

        var conDiaPrimero = DiaMesAnio.Match(texto);
        if (conDiaPrimero.Success)
        {
            return FechaIso8601(
                int.Parse(conDiaPrimero.Groups[3].Value),
                Meses[conDiaPrimero.Groups[2].Value],
                int.Parse(conDiaPrimero.Groups[1].Value));
        }
        return null;
    }
}

File: extraccion/normalizacion.py
"""Convertir el texto leido en un valor con formato, o en nada.

Todo lo de aqui son expresiones regulares y tablas. Ni una sola de estas
funciones adivina: cuando el texto no encaja con el patron devuelven `None`, y
`None` significa «no se pudo leer», que es una respuesta correcta. La respuesta
incorrecta seria un MRN inventado.

Regla permanente 1 del proyecto, aplicada donde mas tienta romperla.
"""

import re

# `DECISIONES.md` §Reglas de formato de los campos.
PATRON_NUMERO_CASO = re.compile(r"\b([A-Z]{4}[0-9]{4})\b")
# ⚠️ El ultimo caracter puede ser una LETRA (`DECISIONES.md`, 2026-09-04). El
# patron viejo pedia once digitos en 3-4-4 y perdia 2 de las 7 cedulas de los
# escaneos reales del dueno: el OCR las leia bien y esta linea las tiraba. Se
# recorto del PDF la banda de la cedula y se miro la imagen: el papel dice `133A`.
#
# La letra NO se toca: ni se sube a mayuscula, ni se cambia por un digito
# «parecido», ni se recorta. Va tal como venia (regla permanente 1).
PATRON_MRN = re.compile(r"\b([0-9]{3})-([0-9]{4})-([0-9]{3}[0-9A-Za-z])\b")
PATRON_UNIDAD = re.compile(r"\b([0-9]{6,})\b")

# Los espacios de sobra que deja el recorte del OCR. Juntar espacios NO es corregir
# lo leido: «Santo  Domingo» y «Santo Domingo» son el mismo texto escrito con la
# basura del recorte, y la regla permanente 1 habla de inventar datos, no de esto.
_ESPACIOS_SEGUIDOS = re.compile(r"\s+")

# Los largos que `unidad_numero` admite, en un solo sitio. `datos/validacion.py`
# y el `CHECK` de `casos` dicen lo mismo (`DECISIONES.md`, 2026-09-02). El patron
# captura 6 o mas digitos a proposito: asi una cifra de 8 se ve entera, se
# reconoce que NO es un numero de unidad y se devuelve None, en vez de recortarle
# los dos ultimos digitos y dar por bueno un numero que no existe.
LARGOS_DE_UNIDAD = frozenset({6, 7})

# Los meses tal como aparecen en estos formularios. Se admite el nombre completo
# y la abreviatura de tres letras, con o sin punto, en los DOS idiomas en que
# llega el formulario: los cuatro documentos de referencia estan en ingles y los
# del dueno en espanol.
_MESES = {
    "enero": 1, "ene": 1,
    "febrero": 2, "feb": 2,
    "marzo": 3, "mar": 3,
    "abril": 4, "abr": 4,
    "mayo": 5, "may": 5,
    "junio": 6, "jun": 6,
    "julio": 7, "jul": 7,
    "agosto": 8, "ago": 8,
    "septiembre": 9, "setiembre": 9,
    "octubre": 10, "oct": 10,
    "noviembre": 11, "nov": 11,
    "diciembre": 12, "dic": 12,
    "january": 1, "jan": 1,
    "february": 2, "feb": 2,
    "march": 3, "mar": 3,
    "april": 4, "apr": 4,
    "may": 5,
    "june": 6, "jun": 6,
    "july": 7, "jul": 7,
    "august": 8, "aug": 8,
    "september": 9, "sept": 9, "sep": 9,  # coinciden con el espanol y valen 9 igual
    "october": 10, "oct": 10,
    "november": 11, "nov": 11,
    "december": 12, "dec": 12,
}

_NOMBRES_DE_MES = "|".join(sorted(_MESES, key=len, reverse=True))
_SUFIJO_ORDINAL = r"(?:st|nd|rd|th)?"

# «September 8, 2026», «Sept-7, 2026», «September 2nd, 2026».
_MES_DIA_ANIO = re.compile(
    rf"\b({_NOMBRES_DE_MES})\.?\s*[-\s]\s*(\d{{1,2}}){_SUFIJO_ORDINAL}\s*,?\s*(\d{{4}})\b",
    re.IGNORECASE,
)
# «8 Sept 2026», «12 September 2026», «14 de marzo de 2027», «14 marzo 2027».
# Los dos «de» del espanol son opcionales: el OCR se come uno a menudo.
_DIA_MES_ANIO = re.compile(
    rf"\b(\d{{1,2}}){_SUFIJO_ORDINAL}\s+(?:de\s+)?({_NOMBRES_DE_MES})\.?"
    rf"\s*(?:de\s+|,\s*)?(\d{{4}})\b",
    re.IGNORECASE,
)

# La fecha en numeros, que es como la escribe el formulario espanol. Medido el
# 2026-09-03: las siete fechas de los dos PDF del dueno son todas `dd-mm-aaaa`,
# y ninguna trae el mes en letras.
#
# El separador puede ser guion, barra o punto. Los tres se ven en escaneos.
_FECHA_ISO = re.compile(r"\b(\d{4})[-/.](\d{1,2})[-/.](\d{1,2})\b")
_FECHA_EN_NUMEROS = re.compile(r"\b(\d{1,2})[-/.](\d{1,2})[-/.](\d{4})\b")

# Un rango —«September 8-11, 2026», «5th till 12th»— NO es una fecha de viaje: no
# se sabe cual de los dos dias vale. Se detecta para poder rechazarlo a proposito
# en vez de quedarse con el primero por casualidad.
_RANGO = re.compile(
    rf"\b\d{{1,2}}{_SUFIJO_ORDINAL}\s*(?:-|–|to|till|until|through)\s*\d{{1,2}}{_SUFIJO_ORDINAL}\b",
    re.IGNORECASE,
)


def _fecha_iso(anio, mes, dia):
    """La fecha en ISO-8601, o None si el dia no existe en ese mes."""
    from datetime import date

    try:
        return date(int(anio), int(mes), int(dia)).isoformat()
    except ValueError:
        return None


def _dia_y_mes_sin_adivinar(primero, segundo):
    """Cual de los dos numeros es el dia y cual el mes, o `(None, None)`.

    Solo responde cuando los DIGITOS lo deciden: un numero mayor que 12 no puede
    ser un mes, asi que el otro lo es. Cuando los dos son 12 o menos la fecha es
    ambigua de verdad —«05-10-2027» es el 5 de octubre o el 10 de mayo— y esta
    funcion se niega a elegir.

    Suponer «dia-mes porque el dueno es de un pais que lo escribe asi» acertaria
    la mayoria de las veces. Las que fallara mueven una fecha de viaje varios
    meses, y es la fecha que decide si un caso sale avisado a tiempo. Regla
    permanente 1: el sistema no adivina.
    """
    if primero > 12 and segundo <= 12:
        return primero, segundo
    if segundo > 12 and primero <= 12:
        return segundo, primero
    return None, None


def _fecha_en_numeros(texto):
    """La fecha escrita en cifras, o None si no hay exactamente UNA sin ambiguedad.

    Se juntan todas las que aparezcan y solo vale si todas dicen lo mismo: dos
    fechas distintas en la misma banda son un rango, y un rango no dice cuando se
    viaja.
    """
    encontradas = set()
    for anio, mes, dia in _FECHA_ISO.findall(texto):
        encontradas.add(_fecha_iso(anio, mes, dia))
    for primero, segundo, anio in _FECHA_EN_NUMEROS.findall(texto):
        dia, mes = _dia_y_mes_sin_adivinar(int(primero), int(segundo))
        if dia is None:
            return None
        encontradas.add(_fecha_iso(anio, mes, dia))
    encontradas.discard(None)
    return encontradas.pop() if len(encontradas) == 1 else None


def normalizar_fecha(texto):
    """«8 Sept 2026» y «08-09-2026» -> «2026-09-08». None si no hay UNA fecha clara.

    Un rango de fechas devuelve None a proposito: en «September 8-11, 2026» no
    hay forma de saber si se viaja el 8 o el 11, y elegir uno seria inventar.

    Las cifras se miran ANTES que el rango, y no es un detalle de orden: el
    detector de rangos casa con «14-03-2027» —dos numeros de dos cifras con un
    guion en medio— y sin esto se comeria todas las fechas del formulario
    espanol dandolas por rangos.
    """
    if not texto:
        return None
    en_numeros = _fecha_en_numeros(texto)
    if en_numeros is not None:
        return en_numeros
    if _RANGO.search(texto):
        return None
    coincidencia = _MES_DIA_ANIO.search(texto)
    if coincidencia is not None:
        mes, dia, anio = coincidencia.groups()
        return _fecha_iso(anio, _MESES[mes.lower()], dia)
    coincidencia = _DIA_MES_ANIO.search(texto)
    if coincidencia is not None:
        dia, mes, anio = coincidencia.groups()
        return _fecha_iso(anio, _MESES[mes.lower()], dia)
    return None


def normalizar_numero_caso(texto):
    """4 letras mayusculas + 4 digitos, o None.

    No se pasa a mayusculas lo que venia en minusculas: si el OCR leyo `casp`
    puede haber leido mal tambien las cifras, y forzarlo a `CASP` esconderia el
    problema en vez de mandarlo a revision.
    """
    if not texto:
        return None
    coincidencia = PATRON_NUMERO_CASO.search(texto)
    return coincidencia.group(1) if coincidencia is not None else None


def normalizar_mrn(texto):
    """3 digitos, guion, 4 digitos, guion, 3 digitos y un caracter mas. O None.

    Ese ultimo caracter puede ser un digito o una LETRA: `055-1111-3853` y
    `066-2222-133A` son las dos FORMAS, y las dos salen de escaneos reales del
    dueno. Palabras suyas el 2026-09-04: «muchas cedulas de miembro tienen una A u
    otra letra al final».

    ⚠️ Los dos valores de arriba van SUSTITUIDOS desde el 2026-09-07: lo que salio
    del papel fue la forma, y la forma esta entera; el ejemplar concreto ya no vive
    en el repositorio. Ver `EN-CURSO.md`, «Los datos de personas reales salen del
    repositorio».

    ⚠️ **Hasta esa fecha esta funcion devolvia None para la segunda forma**, y esa
    era la razon de que 2 de las 7 cedulas del dueno acabaran con el campo vacio:
    el OCR las leia bien y aqui se tiraban.

    Se exigen los guiones tal como estan y la letra se devuelve tal como venia: no
    se sube a mayuscula, no se cambia por un digito «parecido» y no se recorta.
    Cualquiera de las tres cosas es inventar un dato (regla permanente 1), y un MRN
    inventado manda a una persona al templo con la recomendacion equivocada.
    """
    if not texto:
        return None
    coincidencia = PATRON_MRN.search(texto)
    if coincidencia is None:
        return None
    return "-".join(coincidencia.groups())


def normalizar_unidad(texto):
    """Devuelve `(numero_de_unidad, resto_del_texto)`; el numero puede ser None.

    `unidad_numero` son **6 o 7 digitos** segun `DECISIONES.md` (2026-09-02,
    «manda el papel»), segun `validar_unidad_numero` y segun el `CHECK` de la
    tabla `casos` desde la version 2 del esquema. Una cifra de otro largo NO se
    recorta ni se rellena: vuelve como None y el texto crudo viaja aparte para
    que se vea que habia algo.

    El largo admitido se lee de `LARGOS_DE_UNIDAD` para que este archivo y
    `datos/validacion.py` no puedan volver a discrepar en silencio, que es lo
    que paso hasta hoy: el validador aceptaba siete digitos y aqui se tiraban.
    """
    if not texto:
        return None, None
    coincidencia = PATRON_UNIDAD.search(texto)
    if coincidencia is None:
        return None, texto.strip() or None
    digitos = coincidencia.group(1)
    nombre = (texto[: coincidencia.start()] + texto[coincidencia.end() :]).strip(" -–\t")
    numero = digitos if len(digitos) in LARGOS_DE_UNIDAD else None
    return numero, (nombre or None)


def normalizar_numero_de_unidad(texto):
    """Solo el numero de unidad de un texto, o None. Atajo de `normalizar_unidad`.

    Existe aparte porque el numero y el nombre de la unidad se corrigen por
    separado en los formularios reales: hay anotaciones que reescriben el nombre
    y dejan el numero como estaba.
    """
    numero, _ = normalizar_unidad(texto)
    return numero


def normalizar_nombre_de_unidad(texto):
    """Solo el nombre de la unidad, sin su numero, o None."""
    _, nombre = normalizar_unidad(texto)
    return nombre


def normalizar_nombre_del_templo(texto):
    """El nombre del templo tal como esta escrito, con los espacios juntados.

    **No hay catalogo y no se corrige nada** (regla permanente 1, y
    `DECISIONES.md` 2026-09-03: «se lee del papel a `casos.templo_nombre`, texto,
    sin catalogo»). No se compara contra una lista de templos ni se elige «el mas
    parecido»: el catalogo con sus colores es una decision del dueno que todavia no
    ha tomado, y hasta entonces un templo mal leido tiene que verse mal leido y no
    convertido en otro que si esta en la lista.

    Lo unico que se hace es lo mismo que con cualquier otro campo de texto: recortar
    los espacios de los extremos y juntar los de dentro, porque el OCR devuelve «
    Santo  Domingo » y eso es el mismo templo escrito con la basura del recorte. Un
    texto que se queda vacio vuelve None, que significa «no se leyo».
    """
    if not texto:
        return None
    return _ESPACIOS_SEGUIDOS.sub(" ", str(texto)).strip() or None
